Support devolvable widgets

lilyspiniolas opened the pull request:
#### 3ab8a97346bf39e21ca83e1c2b79354baafa3dd7
<pre>
Support devolvable widgets
<a href="https://bugs.webkit.org/show_bug.cgi?id=243899">https://bugs.webkit.org/show_bug.cgi?id=243899</a>
<a href="https://rdar.apple.com/134273374">rdar://134273374</a>

Reviewed by Antti Koivisto and Abrar Rahman Protyasha.

CSS properties which should cause devolvable widgets to show a primitive appearance
instead of a native appearance now do so
(<a href="https://drafts.csswg.org/css-ui-4/#appearance-disabling-properties).">https://drafts.csswg.org/css-ui-4/#appearance-disabling-properties).</a> Layout tests
which used these properties on devolvable widgets have been updated to account for
the new change in appearance. These changes are based on Vitaly Dyachkov&apos;s
(@obyknovenius) earlier proposal.

Test and test result changes:

LayoutTests/fast/forms/auto-fill-button/input-auto-filled-and-obscured.html:
The expected sets the background-color to match the autofill appearance,
but this causes primitive appearance now. The test has been updated to use
primitive appearance from the start to match the expected case.

LayoutTests/fast/forms/hidpi-textfield-background-bleeding.html: Removed.
LayoutTests/platform/mac/fast/forms/hidpi-textfield-background-bleeding-expected.html: Removed.

This test was originally added to detect regressions for a bug where
textfields with background-colors may show the color overflowing due to
AppKit behavior, but now platform/native appearance is disabled when
background color is set, so this issue is no longer possible.

LayoutTests/fast/forms/input-background-ua-media-query-expected.html:
LayoutTests/fast/forms/input-background-ua-media-query.html:

These tests ensure that an input with a parent who has a background color set
does not cause the input to use that background color. It does this by
comparing an input with a parent with a background color to an input
with background color explicitly set to white, which now causes devolution.
The test and the expected have been updated to explicitly use primitive appearance.

LayoutTests/imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-no-fallback-props-001.html:

This test has been upddated to include the longhands for background-position
in the list of styles that should cause primtive appearance  since getComputedStyles
doesn&apos;t return the shorthands.

LayoutTests/platform/ios/fast/forms/select-background-none-expected.txt:

Previously, &apos;background: none&apos; caused the select element to display as
if it had &apos;appearance: none&apos; on iOS. This is no longer the case; although
style &apos;background&apos; causes native appearance to be disabled, we still want
to show the dropdown arrow unless &apos;appearance:none&apos; is explicitly set.
This matches the behavior of Chrome and Firefox.

LayoutTests/platform/ios/fast/forms/select-style-expected.txt:
LayoutTests/platform/mac/fast/forms/select-style-expected.png:
LayoutTests/platform/mac/fast/forms/select-style-expected.txt:

These tests all check the appearances of select elements when they
are styled with background. Previously, background-image:none would
not cause devolution since it matched the user agent style, but since
this is an author level style and background-image causes devolution,
a primtive appearance is used.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/auto-fill-button/input-auto-filled-and-obscured.html:
* LayoutTests/fast/forms/hidpi-textfield-background-bleeding-expected.html: Removed.
* LayoutTests/fast/forms/hidpi-textfield-background-bleeding.html: Removed.
* LayoutTests/fast/forms/input-background-ua-media-query-expected.html:
* LayoutTests/fast/forms/input-background-ua-media-query.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-no-fallback-props-001.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/fast/forms/select-background-none-expected.txt:
* LayoutTests/platform/ios/fast/forms/select-style-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/fast/forms/hidpi-textfield-background-bleeding-expected.html: Removed.
* LayoutTests/platform/mac/fast/forms/select-style-expected.png:
* LayoutTests/platform/mac/fast/forms/select-style-expected.txt:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/process-css-properties.py:
(StylePropertyCodeGenProperties):
(GenerateCSSPropertyNames):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isDevolvableWidget const):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isDevolvableWidget const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLMeterElement.h:
* Source/WebCore/html/HTMLProgressElement.h:
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::devolvableWidgetsEnabledAndSupported):
(WebCore::RenderTheme::adjustStyle):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::nativeAppearanceDisabled const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setNativeAppearanceDisabled):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::setPropertyInternal):
* Source/WebCore/style/PropertyCascade.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomPropertyImpl):
(WebCore::Style::Builder::applyCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeProperty):
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilder.h:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
(JSONCSSPropertiesChecker.check_codegen_properties):

Canonical link: <a href="https://commits.webkit.org/287702@main">https://commits.webkit.org/287702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/196bf1a5dd385272daff29d058f93d16de7d0964

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80533 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34432 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85054 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31515 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82644 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7845 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62923 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20727 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83602 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53029 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73323 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43228 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/79990 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50331 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27481 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29974 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73513 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71476 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28004 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86488 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79592 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7759 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5484 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71215 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7934 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69160 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70455 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17560 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14463 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13412 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101999 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7721 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24839 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7560 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11079 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9365 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -190,6 +190,8 @@ static bool isAppearanceAllowedForAllElements(StyleAppearance appearance)
 void RenderTheme::adjustStyle(RenderStyle& style, const Element* element, const RenderStyle* userAgentAppearanceStyle)
 {
+    (void)userAgentAppearanceStyle;
+
```
Nit - I think you can remove this?

jesxilin commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -2103,6 +2103,25 @@ bool HTMLInputElement::isEmptyValue() const
+        || m_inputType->isNumberField()
+        || m_inputType->isColorControl()
+        || m_inputType->isTextButton();
+    }
```
I think there's an indentation here

pxlcoder commented on LayoutTests/fast/forms/auto-fill-button/input-auto-filled-and-obscured.html:
```diff
@@ -1,5 +1,10 @@
 <html>
+<style>
+    input {
+        appearance:none;
```
Nit: Add a space before `none`.

pxlcoder commented on LayoutTests/fast/forms/input-background-ua-media-query-expected.html:
```diff
@@ -1,4 +1,4 @@
 body { background-color: green }
 </style>
-<input value=foo style="background-color:white">
+<input value=foo style="border: 2px gray inset; background-color:white">
```
Ideally each test change would be individually justified/explained in detail in the commit message.

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -2103,6 +2103,25 @@ bool HTMLInputElement::isEmptyValue() const
+    return m_inputType->isTextType()
+        || m_inputType->isTelephoneField()
+        || m_inputType->isURLField()
+        || m_inputType->isURLField()
```
The “URL field” check is duplicated.

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -2103,6 +2103,25 @@ bool HTMLInputElement::isEmptyValue() const
+        || m_inputType->isSearchField()
+        || m_inputType->isPasswordField()
+        || m_inputType->isDateField()
+        || m_inputType->isMonthField()
```
Not at a laptop right now, but I think there might be a better way to do this, which groups related input types.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -1794,6 +1799,7 @@
                 "name-for-methods": "Clip",
                 "fill-layer-property": true,
                 "separator": " ",
+                 "disables-native-appearance": true,
```
```suggestion
                "disables-native-appearance": true,
```

nt1m commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -205,28 +207,40 @@ void RenderTheme::adjustStyle(RenderStyle& style, const Element* element, const
-        style.setUsedAppearance(appearance);
-    }
+    if (hasDropdownIndicator) {
+        if (userAgentAppearanceStyle && isControlStyled(style, *userAgentAppearanceStyle)) {
```
You can entirely get rid of `isControlStyled` and inline most of the relevant logic here (e.g. only the select cases)

nt1m reviewed (commented):
Can we see the progressions in TestExpectations?

pxlcoder reviewed (commented):
We need to preserve the old behavior with a new `SDKAlignedBehavior` in `RuntimeApplicationChecksCocoa`.

nt1m commented on Source/WebCore/dom/Element.h:
```diff
@@ -604,6 +604,8 @@ class Element : public ContainerNode {
     virtual bool isSliderThumbElement() const { return false; }
     virtual bool isHTMLTablePartElement() const { return false; }
 
+    virtual bool isDevolvableWidget() const { return false; }
```
virtual functions might be slow, worth A/B testing perf benchmarks (especially Speedometer)

anttijk commented on Source/WebCore/style/PropertyCascade.cpp:
```diff
@@ -106,11 +106,18 @@ void PropertyCascade::setPropertyInternal(Property& property, CSSPropertyID id,
 
     if (matchedProperties.linkMatchType == SelectorChecker::MatchAll) {
-        property.cssValue[0] = &cssValue;
+        property.cascadeLevels[SelectorChecker::MatchDefault] = cascadeLevel;
+        property.cssValue[SelectorChecker::MatchDefault] = &cssValue;
+
+        property.cascadeLevels[SelectorChecker::MatchLink] = cascadeLevel;
         property.cssValue[SelectorChecker::MatchLink] = &cssValue;
+
+        property.cascadeLevels[SelectorChecker::MatchVisited] = cascadeLevel;
         property.cssValue[SelectorChecker::MatchVisited] = &cssValue;
```
I don't expect perf impact but it is good to check the perf bots at least after landing. This is pretty hot code after all.

anttijk commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -398,6 +398,9 @@ void Builder::applyProperty(CSSPropertyID id, CSSValue& value, SelectorChecker::
+    if (!isUnset && cascadeLevel == CascadeLevel::Author && m_state.element()->isDevolvableWidget() && CSSProperty::disablesNativeAppearance(id) && m_state.applyPropertyToRegularStyle())
+        style.setNativeAppearanceDisabled(true);
+
     BuilderGenerated::applyProperty(id, m_state, valueToApply.get(), valueType);
```
This extra adjustment would look nicer after the`BuilderGenerated::applyProperty` call since that is still the main thing happening here.

lilyspiniolas committed:
Support devolvable widgets
https://bugs.webkit.org/show_bug.cgi?id=243899
rdar://134273374

Reviewed by Antti Koivisto and Abrar Rahman Protyasha.

CSS properties which should cause devolvable widgets to show a primitive appearance
instead of a native appearance now do so
(https://drafts.csswg.org/css-ui-4/#appearance-disabling-properties). Layout tests
which used these properties on devolvable widgets have been updated to account for
the new change in appearance. These changes are based on Vitaly Dyachkov's
(@obyknovenius) earlier proposal.

Test and test result changes:

LayoutTests/fast/forms/auto-fill-button/input-auto-filled-and-obscured.html:
The expected sets the background-color to match the autofill appearance,
but this causes primitive appearance now. The test has been updated to use
primitive appearance from the start to match the expected case.

LayoutTests/fast/forms/hidpi-textfield-background-bleeding.html: Removed.
LayoutTests/platform/mac/fast/forms/hidpi-textfield-background-bleeding-expected.html: Removed.

This test was originally added to detect regressions for a bug where
textfields with background-colors may show the color overflowing due to
AppKit behavior, but now platform/native appearance is disabled when
background color is set, so this issue is no longer possible.

LayoutTests/fast/forms/input-background-ua-media-query-expected.html:
LayoutTests/fast/forms/input-background-ua-media-query.html:

These tests ensure that an input with a parent who has a background color set
does not cause the input to use that background color. It does this by
comparing an input with a parent with a background color to an input
with background color explicitly set to white, which now causes devolution.
The test and the expected have been updated to explicitly use primitive appearance.

LayoutTests/imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-no-fallback-props-001.html:

This test has been upddated to include the longhands for background-position
in the list of styles that should cause primtive appearance  since getComputedStyles
doesn't return the shorthands.

LayoutTests/platform/ios/fast/forms/select-background-none-expected.txt:

Previously, 'background: none' caused the select element to display as
if it had 'appearance: none' on iOS. This is no longer the case; although
style 'background' causes native appearance to be disabled, we still want
to show the dropdown arrow unless 'appearance:none' is explicitly set.
This matches the behavior of Chrome and Firefox.

LayoutTests/platform/ios/fast/forms/select-style-expected.txt:
LayoutTests/platform/mac/fast/forms/select-style-expected.png:
LayoutTests/platform/mac/fast/forms/select-style-expected.txt:

These tests all check the appearances of select elements when they
are styled with background. Previously, background-image:none would
not cause devolution since it matched the user agent style, but since
this is an author level style and background-image causes devolution,
a primtive appearance is used.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/auto-fill-button/input-auto-filled-and-obscured.html:
* LayoutTests/fast/forms/hidpi-textfield-background-bleeding-expected.html: Removed.
* LayoutTests/fast/forms/hidpi-textfield-background-bleeding.html: Removed.
* LayoutTests/fast/forms/input-background-ua-media-query-expected.html:
* LayoutTests/fast/forms/input-background-ua-media-query.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-no-fallback-props-001.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/fast/forms/select-background-none-expected.txt:
* LayoutTests/platform/ios/fast/forms/select-style-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/fast/forms/hidpi-textfield-background-bleeding-expected.html: Removed.
* LayoutTests/platform/mac/fast/forms/select-style-expected.png:
* LayoutTests/platform/mac/fast/forms/select-style-expected.txt:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/process-css-properties.py:
(StylePropertyCodeGenProperties):
(GenerateCSSPropertyNames):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isDevolvableWidget const):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isDevolvableWidget const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLMeterElement.h:
* Source/WebCore/html/HTMLProgressElement.h:
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::devolvableWidgetsEnabledAndSupported):
(WebCore::RenderTheme::adjustStyle):
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::nativeAppearanceDisabled const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setNativeAppearanceDisabled):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::setPropertyInternal):
* Source/WebCore/style/PropertyCascade.h:
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomPropertyImpl):
(WebCore::Style::Builder::applyCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeProperty):
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilder.h:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
(JSONCSSPropertiesChecker.check_codegen_properties):

Canonical link: https://commits.webkit.org/287702@main
